mono/mini/method-to-ir.c

```diff
@@ -2230,41 +2241,35 @@ test_tailcall (MonoCompile *cfg, MonoBoolean tailcall)
 	// Do not change "tailcalllog" here without changing other places, e.g. tests that search for it.
 	//
 	g_assertf (tailcall || !mini_get_debug_options ()->test_tailcall_require, "tailcalllog fail from %s", cfg->method->name);
-	mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_TAILCALL, "tailcalllog %s from %s", tailcall ? "success" : "fail", cfg->method->name);
+	tailcall_print ("tailcalllog %s from %s", tailcall ? "success" : "fail", cfg->method->name);
 }
 
-inline static MonoCallInst *
+static MonoCallInst *
 mono_emit_call_args (MonoCompile *cfg, MonoMethodSignature *sig, 
-					 MonoInst **args, int calli, int virtual_, int tail, int rgctx, int unbox_trampoline, MonoMethod *target)
+		     MonoInst **args, gboolean calli, gboolean virtual_, gboolean tailcall,
+		     gboolean rgctx, gboolean unbox_trampoline, MonoMethod *target)
 {
 	MonoType *sig_ret;
 	MonoCallInst *call;
-#ifdef MONO_ARCH_SOFT_FLOAT_FALLBACK
-	int i;
-#endif
 
-	if (cfg->llvm_only) {
-		if (tail)
-			test_tailcall (cfg, FALSE);
-		tail = FALSE;
-	}
+	g_assert (!cfg->llvm_only || !tailcall); // FIXME; duplicated in is_supported_tail_call
 
-	if (tail && (debug_tailcall_break_compile || debug_tailcall_break_run)
+	if (tailcall && (debug_tailcall_break_compile || debug_tailcall_break_run)
 		&& mono_is_usermode_native_debugger_present ()) {
 
 		if (debug_tailcall_break_compile)
 			G_BREAKPOINT ();
 
-		if (tail && debug_tailcall_break_run) { // Can change tail in debugger.
+		if (tailcall && debug_tailcall_break_run) { // Can change tailcall in debugger.
 			MonoInst *brk;
 			MONO_INST_NEW (cfg, brk, OP_BREAK);
 			MONO_ADD_INS (cfg->cbb, brk);
 		}
 	}
 
-	if (tail) {
+	if (tailcall) {
 		mini_profiler_emit_tail_call (cfg, target);
-		MONO_INST_NEW_CALL (cfg, call, virtual_ ? OP_TAILCALL_MEMBASE : OP_TAILCALL);
+		MONO_INST_NEW_CALL (cfg, call, calli ? OP_TAILCALL_REG : virtual_ ? OP_TAILCALL_MEMBASE : OP_TAILCALL);
 	} else
```

Comment:
This could use a parentheses.

Reply:
It works intuitively but if you insist..

Reply:
Is the pinvoke_callconv thing calli-specific? It is messier to keep it so but I can.